src/architecture/unixfs.md

```diff
@@ -0,0 +1,618 @@
+
+## Design Decision Rationale
+
+### `mtime` and `mode` Metadata Support in UnixFSv1.5
+
+Metadata support in UnixFSv1.5 has been expanded to increase the number of possible
+use cases. These include rsync and filesystem based package managers.
+
+Several metadata systems were evaluated, as discussed in the following sections.
+
+#### Separate Metadata Node
+
+In this scheme, the existing `Metadata` message is expanded to include additional
+metadata types (`mtime`, `mode`, etc). It contains links to the actual file data
+but never the file data itself.
+
+This was ultimately rejected for a number of reasons:
+
+1. You would always need to retrieve an additional node to access file data which
+  limits the kind of optimizations that are possible. For example many files are
+  under the 256 KiB block size limit, so we tend to inline them into the describing
+  UnixFS `File` node. This would not be possible with an intermediate `Metadata` node.
+2. The `File` node already contains some metadata (e.g. the file size) so metadata
+  would be stored in multiple places which complicates forwards compatibility with
+  UnixFSv2 as to map between metadata formats potentially requires multiple fetch
+  operations.
+
+#### Metadata in the Directory
+
+Repeated `Metadata` messages are added to UnixFS `Directory` and `HAMTShard` nodes,
+the index of which indicates which entry they are to be applied to. Where entries are
+`HAMTShard`s, an empty message is added.
+
+One advantage of this method is that if we expand stored metadata to include entry
+types and sizes we can perform directory listings without needing to fetch further
+entry nodes (excepting `HAMTShard` nodes), though without removing the storage of
+these datums elsewhere in the spec we run the risk of having non-canonical data
+locations and perhaps conflicting data as we traverse through trees containing
+both UnixFS v1 and v1.5 nodes.
+
+This was rejected for the following reasons:
+
+1. When creating a UnixFS node there's no way to record metadata without wrapping
+  it in a directory.
+2. If you access any UnixFS node directly by its [CID], there is no way of recreating
+   the metadata which limits flexibility.
+3. In order to list the contents of a directory including entry types and sizes,
+   you have to fetch the root node of each entry anyway so the performance benefit
+   of including some metadata in the containing directory is negligible in this
+   use case.
+
+#### Metadata in the File
+
+This adds new fields to the UnixFS `Data` message to represent the various metadata fields.
+
+It has the advantage of being simple to implement, metadata is maintained whether
+the file is accessed directly via its [CID] or via an IPFS path that includes a
+containing directory, and by keeping the metadata small enough we can inline root
+UnixFS nodes into their CIDs so we can end up fetching the same number of nodes if
+we decide to keep file data in a leaf node for deduplication reasons.
```

Comment:
```suggestion
It has the advantage of being simple to implement. Metadata is maintained whether
the file is accessed directly via its [CID] or via an IPFS path that includes a
containing directory. In addition, metadata is kept small enough that we can inline root
UnixFS nodes into their CIDs so that we can end up fetching the same number of nodes if
we decide to keep file data in a leaf node for deduplication reasons.
```